packages/hardhat-core/src/internal/vendor/await-semaphore/index.ts

```diff
@@ -64,3 +67,154 @@ export class Mutex extends Semaphore {
+  private _mutexFilePath: string;
+  private _timeoutInMs: number;
+  private _defaultTimeoutInMs = 20000;
+  private _safeMarginInMs = 2000;
+  private _mutexReleaseWaitingTimeInMs = 500;
+
+  constructor(mutexName: string, timeoutInMs?: number) {
+    this._log = debug("hardhat:await-semaphore:multi-process-mutex");
+
+    this._log(`Creating mutex with name '${mutexName}'`);
+
+    this._mutexFilePath = `${os.tmpdir()}/${mutexName}.txt`;
+    this._timeoutInMs = timeoutInMs ?? this._defaultTimeoutInMs;
+  }
+
+  public async use<T>(f: () => Promise<T>): Promise<T> {
+    this._log(
+      `Starting mutex process withy mutex file '${this._mutexFilePath}'`
+    );
+
+    while (true) {
+      if (await this._tryToAcquireMutex()) {
+        // Mutex has been acquired
+        return this._executeFunctionAndReleaseMutex(f);
+      }
+
+      // Mutex not acquired
+      if (this._isMutexFileTooOld()) {
+        // If the mutex file is too old, it likely indicates a deadlock, so the file should be removed
+        this._log(
+          `Current mutex file is too old, removing it at path '${this._mutexFilePath}'`
+        );
+        this._deleteMutexFile();
+      } else {
+        // wait
+        await this._waitMs();
+      }
+    }
+  }
+
+  private async _tryToAcquireMutex() {
+    try {
+      // Create a file only if it does not exist
+      fs.writeFileSync(this._mutexFilePath, "", { flag: "wx+" });
+      return true;
+    } catch (error: any) {
+      if (error.code === "EEXIST") {
+        // File already exists, so the mutex is already acquired
+        return false;
+      }
+
+      throw error;
+    }
+  }
+
+  private async _executeFunctionAndReleaseMutex<T>(
+    f: () => Promise<T>
+  ): Promise<T> {
+    this._log(`Mutex acquired at path '${this._mutexFilePath}'`);
+
+    // Set a timeout to ensure that the function does not take longer than the allowed time
+    let timeoutHandle: NodeJS.Timeout;
+    const timeoutChecker = new Promise<never>((_, reject) => {
+      timeoutHandle = setTimeout(() => {
+        reject(
+          new Error(
+            "The function's execution timed out while it had ownership of the mutex"
+          )
+        );
+      }, this._defaultTimeoutInMs);
+    });
+
+    try {
+      // Execute the function passed to the mutex, no other function will be able to run at the same time.
+      // If the function will take more than the timeout, it will be terminated with an error and the mutex will be released.
+      // Promise.race is used to race the timeoutChecker against the execution of fn().
+      // Whichever completes first will be the outcome of Promise.race.
+      return await Promise.race([f(), timeoutChecker]).then((result) => {
+        clearTimeout(timeoutHandle);
+
+        // Release the mutex
+        this._deleteMutexFile();
+
+        this._log(`Mutex released at path '${this._mutexFilePath}'`);
+
+        return result;
+      });
+    } catch (error: any) {
```

Comment:
Perhaps this catch and the `deleteMutexFile` call in line 165 could be replaced just with a `finally { this._deleteMutexFile() }`? I'm not 100% sure, but I think it's equivalent and it would better convey the intended behavior.

Reply:
This logic has been removed, see previous comments